binlog_reconciler.go

```diff
@@ -0,0 +1,223 @@
+package ghostferry
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/siddontang/go-mysql/schema"
+	"github.com/sirupsen/logrus"
+)
+
+// This event generates a REPLACE INTO SQL statement to overwrite a row if an
+// INSERT or UPDATE occured. For a DELETE statement, it generates the
+// corresponding DELETE statement based on the primary key.
+//
+// This is used during the resume when we are catching up to the binlog events
+// missed by Ghostferry while it is down.
+type ReconcilationDMLEvent struct {
+	*DMLEventBase
```

Comment:
Why do you need to embed this? It doesn't make sense to call this struct a `*DMLEvent` either. It's not really an event itself. You may store the event that produced this as a field on the struct but calling this a DML event itself is very confusing

Is the only goal of this struct to provide a different `AsSQLString` so that it works as-is with the BinlogWriter? I feel like we need to refactor the current `BinlogWriter` to become a `MultiStatementWriter`.


```
type MultiStatementWriter struct{}  // current BinlogWriter implementation

type BinlogWriter struct{
  writer *MultiStatementWriter
}

type BinlogReconciler struct{
  writer *MultiStatementWriter
}
```

Reply:
I tried to refactor the BinlogWriter and it turned out to be a bit more difficult and tedious than I would have liked. Since the interface of the DMLEvent is mainly to call `AsSQLString`, I decided that it's easier to work with this version and refactor later.